docs/lang/articles/advanced/sparse.md

```diff
@@ -2,316 +2,281 @@
 
-One caveat of this design is that voxels in the same `dense` block can no longer change their activity flexibly.
-Instead, they share a single activity flag. To address this issue,
-the `bitmasked` SNode additionally allocates 1-bit per voxel data to represent the voxel activity.
+:::note
+Reading an inactive pixel returns zero.
+:::
 
-### A typical sparse data structure
+### Pointer SNode
 
-Sparse data structures in Taichi are usually composed of `pointer`, `dense`, and `bitmasked` SNodes.
-The code snippet below creates an 8x8 sparse grid, with the top level being 4x4 pointer arrays,
-and each pointer points to a 2x2 dense block.
+```python {2} title=pointer.py
+x = ti.field(ti.f32)
+block = ti.root.pointer(ti.ij, (4,4))
+pixel = block.dense(ti.ij, (2,2))
+pixel.place(x)
 
-```python
-x = ti.field(dtype=ti.i32)
+@ti.kernel
+def activate():
+    x[2,3] = 1.0
+    x[2,4] = 2.0
 
-block = ti.root.pointer(ti.ij, (4, 4))
-pixel = block.dense(ti.ij, (2, 2))
-pixel.place(x)
+@ti.kernel
+def print_active():
+    for i, j in block:
+        print("Active block", i, j)
+    # output: Active block 1 1
+    #         Active block 1 2
+    for i, j in x:
+        print('field x[{}, {}] = {}'.format(i, j, x[i, j]))
+    # output: field x[2, 2] = 0.000000
+    #         field x[2, 3] = 1.000000
+    #         field x[3, 2] = 0.000000
+    #         field x[3, 3] = 0.000000
+    #         field x[2, 4] = 2.000000
+    #         field x[2, 5] = 0.000000
+    #         field x[3, 4] = 0.000000
+    #         field x[3, 5] = 0.000000
 ```
+The code snippet above creates an 8x8 sparse grid, with the top-level being 4x4 pointer arrays (line 2 of `pointer.py`),
+and each pointer points to a 2x2 dense block. The `activate` function could automatically activates `block[1,1]` which includes `x[2,3]` and `block[1,2]` which includes `x[2, 4]`. Other pixels of `block[1,1]` (`x[2,2], x[3, 2], x[3, 3]`) and `block[1,2]` (`x[2, 5], x[3, 4], x[3, 5]`) are also implicitly activated if `pixel` is a dense node. Because all pixels in the dense block share the same activity value.
+As the below figure shows, the active blocks and pixels are green. You can write and read the sparse field like a dense field using indices.
 
-![image](https://raw.githubusercontent.com/taichi-dev/public_files/master/taichi/doc/sparse_grids_2d.png)
+<center>
 
-## Computation on sparse data structures
+![Pointer](https://raw.githubusercontent.com/FantasyVR/public_files/sparse_computation/taichi/doc/pointer_small.png)
 
-### Activation on write
+</center>
 
-When writing to an inactive cell on a sparse data structure, Taichi automatically populates the data structure.
 
```

Comment:
```suggestion
Executing the `activate()` function automatically activates `block[1,1]`, which includes `x[2,3]`, and `block[1,2]`, which includes `x[2,4]`. Other pixels of `block[1,1]` (`x[2,2], x[3,2], x[3,3]`) and `block[1,2]` (`x[2,5], x[3,4], x[3,5]`) are also implicitly activated because all pixels in the dense block share the same activity value.

```